src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala

```diff
@@ -532,41 +584,103 @@ case class Block(center: Array[Double], faces: Seq[Face], generation: Int=0) ext
         allVertices.map { v => v(2) }.sum / allVertices.size
       )
     } else {
-      Array(
-        // Factor of 3 comes from: centroid / (2.0 * (volume/6.0))
-        3.0 * centroidX / totalVolume,
-        3.0 * centroidY / totalVolume,
-        3.0 * centroidZ / totalVolume
-      )
+      val increments: Seq[(Double, Double, Double)] = faces.flatMap { face =>
+        vertices(face).map { faceVertex =>
+          val vertexIndex = vertices(face).indexOf(faceVertex)
```

Comment:
I think you can avoid this `indexOf` call, which I could imagine getting a bit expensive, with something like this:
```scala
vertices(face).zipWithIndex.map { case (index, faceVertex) =>
    ...
}
```